### PR TITLE
fix octree bounds reducing

### PR DESCRIPTION
**Pull request by pkuhto**
for some resolution/bounds ratio octree bounds can be unexpectedly reduced

### CODE REVIEW

**Comment by jspricke** [comment]
Interesting, can you add a unit test with an example?

@jkammerl can you have a look?

**Comment by pkuhto** [comment]
const double someResolution(10);
	const int someDepth(4);
	const double desiredMax = ((1<<someDepth) + 0.5)*someResolution;
	pcl::octree::OctreePointCloud<PointT> tree(someResolution);
	tree.defineBoundingBox(0, 0, 0, desiredMax, desiredMax, desiredMax);
	double min_x, min_y, min_z, max_x, max_y, max_z;
	tree.getBoundingBox(min_x, min_y, min_z, max_x, max_y, max_z);
	assert(max_x >= desiredMax && max_y >= desiredMax && max_z >= desiredMax);

of course bounds can be corrected in adoptBoundingBoxToPoint with treedepth=srcTreeDepth+2 (worst scenario), but I didn't use bounds correction for perfomance reasons and got points in invalid leaves.

**Comment by jspricke** [comment]
Can you put this into a unit test and add it to the pull request? That would be great :).

**Comment by SergioRAgostinho** [comment]
Nice catch. 👍 

1. Please follow our [coding style guide](http://pointclouds.org/documentation/advanced/pcl_style_guide.php), more specifically notice how the spacing is used before parenthesis. 

2. @jspricke I'm not sure how those ```ref_values``` in the  [GFPFH test](https://github.com/PointCloudLibrary/pcl/blob/master/test/features/test_pfh_estimation.cpp#L540-L579) were originally calculated, but I have the impression they relied on the bugged implementation of the octree, especially considering it's using the bunny pointcloud and not something much more synthetic that could be manually computed. Can we consider to update them?

**Comment by jspricke** [comment]
@SergioRAgostinho I guess they where computed and then later manually verified. If we don't change GFPFH in between, putting in the newly computed values should be fine. @pkuhto can you add that to the pull request as well? Thanks!

**Comment by SergioRAgostinho** [comment]
LGTM 👍  Squash all the commits into a single one and this should be good to go.

File: test/octree/test_octree.cpp [hunk 1]
```diff
@@ -1699,6 +1699,38 @@ TEST (PCL, Octree_Pointcloud_Adjacency)
...
+    tree.getBoundingBox (min_x, min_y, min_z, max_x, max_y, max_z);
+    const unsigned int depth = tree.getTreeDepth ();
+    tree.defineBoundingBox (min_x, min_y, min_z, max_x, max_y, max_z);
+    
+    ASSERT_EQ( depth == tree.getTreeDepth (), true );
+    
+    double min_x2, min_y2, min_z2, max_x2, max_y2, max_z2;
+    tree.getBoundingBox (min_x2, min_y2, min_z2, max_x2, max_y2, max_z2);
+    
+    ASSERT_EQ( min_x2 == min_x, true );
```
Comment (taketwo) [inline test/octree/test_octree.cpp]: Please use [`ASSERT_DOUBLE_EQ`](https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md#floating-point-macros) to make assertions about floating-point numbers.

File: test/octree/test_octree.cpp [hunk 2]
```diff
@@ -1699,6 +1699,38 @@ TEST (PCL, Octree_Pointcloud_Adjacency)
...
+    ASSERT_GE( DESIRED_MIN, min_x );
+    
+    const double LARGE_MIN = 1e7-45*SOME_RESOLUTION;
+    const double LARGE_MAX = 1e7-5*SOME_RESOLUTION;
+    tree.defineBoundingBox (LARGE_MIN, LARGE_MIN, LARGE_MIN, LARGE_MAX, LARGE_MAX, LARGE_MAX);
+    tree.getBoundingBox (min_x, min_y, min_z, max_x, max_y, max_z);
+    const unsigned int depth = tree.getTreeDepth ();
+    tree.defineBoundingBox (min_x, min_y, min_z, max_x, max_y, max_z);
+    
+    ASSERT_EQ( depth == tree.getTreeDepth (), true );
```
Comment (taketwo) [inline test/octree/test_octree.cpp]: Why not simply: `ASSERT_EQ(depth, tree.getTreeDepth ())`?

**Comment by taketwo** [comment]
> Squash all the commits

@SergioRAgostinho note that Github can do this. You will get an option to squash commits before merging when you press "Merge pull request" button.

**Comment by SergioRAgostinho** [comment]
@taketwo  Should we squash-merge this and then just submit a minor corrections to those asserts?

**Comment by taketwo** [comment]
Sure. Please merge and I'll make a correction PR.

**Pull request merged**